cmd/skiff/top.go

```diff
@@ -61,10 +74,65 @@ func (h *FileHeap) Pop() interface{} {
 	return item
 }
 
+// HumanReadableSize converts a byte count to a human readable string
+// From https://yourbasic.org/golang/formatting-byte-size-to-human-readable-format/
+func HumanReadableSize(b int64) string {
+	const unit = 1000
+	if b < unit {
+		return fmt.Sprintf("%d B", b)
+	}
+	div, exp := int64(unit), 0
+	for n := b / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	return fmt.Sprintf("%.1f %cB",
+		float64(b)/float64(div), "kMGTPE"[exp])
+}
+
+// getFilteredLayers returns only the layers that needs to be
+// processed/extracted e.g. after user specifies specific layer(s)
+// using --layer, we shouldn't be processing all the layers.
+func getFilteredLayers(img types.Image, layers []string) ([]types.BlobInfo, error) {
+	allLayers := img.LayerInfos()
+
+	if len(layers) == 0 {
+		return allLayers, nil // No filtering needed
+	}
+
+	var filteredLayers []types.BlobInfo
+	seenLayers := make(map[string]bool)
+
+	for _, filter := range layers {
+		matchedLayers := []types.BlobInfo{}
+
+		for _, layer := range allLayers {
+			if layer.Digest.Encoded() == filter || strings.HasPrefix(layer.Digest.Encoded(), filter) {
+				matchedLayers = append(matchedLayers, layer)
+			}
+		}
+
+		if len(matchedLayers) == 0 {
+			return nil, fmt.Errorf("layer %s not found in image", filter)
+		}
+		if len(matchedLayers) > 1 {
+			return nil, fmt.Errorf("multiple layers match shortened digest %s", filter)
+		}
+
+		layer := matchedLayers[0]
+		if !seenLayers[layer.Digest.String()] {
+			filteredLayers = append(filteredLayers, layer)
+			seenLayers[layer.Digest.String()] = true
+		}
+	}
+
+	return filteredLayers, nil
```

Comment:
How about this? (needs a gofmt pass)

```suggestion
   layerMap := make(map[string]types.BlobInfo)
   for _, layer := range allLayers {
       layerMap[layer.Digest.Encoded()] = layer
   }
   var filteredLayers []types.BlobInfo
   for _, filter := range layers {
       var matches []string
       for k := range layerMap {
           if k == filter || strings.HasPrefix(k, filter) {
               matches = append(matches, k)
           }
       }
       if len(matches) == 0 {
           return nil, fmt.Errorf("layer %s not found in image", filter)
       }
       if len(matches) > 1 {
           return nil, fmt.Errorf("multiple layers match shortened digest %s", filter)
       }
       matchedKey := matches[0]
       filteredLayers = append(filteredLayers, layerMap[matchedKey])
       delete(layerMap, matchedKey) // pop so it can't be reused
   }
   return filteredLayers, nil
```

Reply:
That's much more efficient, thanks!